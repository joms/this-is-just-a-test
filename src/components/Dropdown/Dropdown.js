import React, { useState, useMemo, useEffect } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

const Dropdown = ({ label, onChange, options, placeholder, value, error, hint }) => {
    const [open, setOpen] = useState(false);

    const close = () => setOpen(false);
    const show = () => setOpen(true);

    useEffect(() => {
        if (open) {
            window.addEventListener('click', close);
        } else {
            window.removeEventListener('click', close);
        }

        return () => window.removeEventListener('click', close);
    }, [open]);

    const valueText = useMemo(() => {
        if (!value) return false;
        const v = options.find(p => p.value === value).text;

        if (v) {
            return v;
        }

        return false;
    }, [value, options]);

    // TODO Support arrow key navigation https://github.com/Semantic-Org/Semantic-UI-React/blob/master/src/modules/Dropdown/Dropdown.js#L514

    return (
        <>
            <label>{label}</label>
            <div className={cx('dropdown', { error, open })} onClick={show} onFocus={show} onBlur={close} tabIndex="0">
                <div className={cx({ value: valueText }, { placeholder: !valueText })}>{valueText || placeholder}</div>
                <i className="arrow" />
                <div className="dropdown-options">
                    {options.map(option => (
                        <div
                            className="dropdown-option"
                            onClick={() => onChange(option.value)}
                            key={option.value + option.text}
                        >
                            {option.text}
                        </div>
                    ))}
                </div>
                {error && <span className="error">{error}</span>}
                {hint && <span className="hint">{hint}</span>}
            </div>
        </>
    );
};

Dropdown.defaultProps = {
    value: '',
    placeholder: '',
    error: '',
    hint: '',
};

Dropdown.propTypes = {
    label: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    options: PropTypes.arrayOf(
        PropTypes.shape({
            value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
            text: PropTypes.string,
        })
    ).isRequired,
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    placeholder: PropTypes.string,
    error: PropTypes.string,
    hint: PropTypes.string,
};

export default Dropdown;
