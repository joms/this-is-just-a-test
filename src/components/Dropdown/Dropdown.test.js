import React from 'react';
import { shallow } from 'enzyme';
import Dropdown from '.';

describe('<App>', () => {
    let wrapper;
    let onChange;

    beforeEach(() => {
        onChange = jest.fn();
        wrapper = shallow(
            <Dropdown
                label="Hello, Dropdown!"
                options={[{ value: 3.14, text: 'PI' }, { value: 6.28, text: 'Tau' }]}
                onChange={onChange}
            />
        );
    });

    it('matches snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('opens', () => {
        expect(wrapper.find('.dropdown').hasClass('open')).toBeFalsy();
        wrapper.find('.dropdown').simulate('click');
        expect(wrapper.find('.dropdown').hasClass('open')).toBeTruthy();
    });

    it('handles change', () => {
        wrapper.find('.dropdown').simulate('click');
        wrapper
            .find('.dropdown-option')
            .first()
            .simulate('click');

        expect(onChange).toHaveBeenCalled();
    });
});
