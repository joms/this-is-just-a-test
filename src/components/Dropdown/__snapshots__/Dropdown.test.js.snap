// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App> matches snapshot 1`] = `
<Fragment>
  <label>
    Hello, Dropdown!
  </label>
  <div
    className="dropdown"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    tabIndex="0"
  >
    <div
      className="placeholder"
    />
    <i
      className="arrow"
    />
    <div
      className="dropdown-options"
    >
      <div
        className="dropdown-option"
        key="3.14PI"
        onClick={[Function]}
      >
        PI
      </div>
      <div
        className="dropdown-option"
        key="6.28Tau"
        onClick={[Function]}
      >
        Tau
      </div>
    </div>
  </div>
</Fragment>
`;
